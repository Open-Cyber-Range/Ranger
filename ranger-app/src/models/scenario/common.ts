export type Source = {
  name: string;
  version: string;
};
