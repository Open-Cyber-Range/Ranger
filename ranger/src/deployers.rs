use crate::{
    node::GetCapabilities,
    node::NodeClient,
    services::deployment::{
        initiate_node_clients, AddDeploymentGroups, DeploymentGroup, DeploymentGroupMap,
        DeploymentManager, NodeClientFilter,
    },
    templater::{initiate_template_clients, TemplateClientFilter},
};

use actix::{Actor, Addr, Context, Handler, Message, MessageResponse};
use anyhow::{anyhow, Result};
use futures::future::join_all;
use log::error;
use ranger_grpc::{capabilities::DeployerTypes, Capabilities};
use serde::Serialize;
use std::collections::HashMap;

#[derive(Message, Debug, Clone)]
#[rtype(result = "()")]
pub struct AddDeployerGroups(pub(crate) DeployerGroups);

#[derive(Message, Debug, PartialEq)]
#[rtype(result = "DeployerGroups")]
pub struct GetDeployerGroups;

#[derive(Debug, Default, Clone)]
pub struct Deployer {
    pub deployer_name: String,
    pub deployer_ip: String,
    pub capabilities: Option<Capabilities>,
}

#[derive(Debug, Default, Serialize, Clone)]
pub struct DeployerGroup {
    pub machiners: HashMap<String, String>,
    pub switchers: HashMap<String, String>,
    pub templaters: HashMap<String, String>,
}

impl DeployerGroup {
    pub async fn start(&self) -> DeploymentGroup {
        let machiners = join_all(initiate_node_clients(self.machiners.clone())).await;
        let switchers = join_all(initiate_node_clients(self.switchers.clone())).await;
        let templaters = join_all(initiate_template_clients(self.templaters.clone())).await;
        DeploymentGroup {
            machiners: machiners.filter_node_clients(),
            switchers: switchers.filter_node_clients(),
            templaters: templaters.filter_template_clients(),
        }
    }

    pub fn insert_by_capability(&mut self, deployer: &Deployer) {
        if let Some(capabilities) = &deployer.capabilities {
            capabilities
                .values()
                .into_iter()
                .for_each(|capability| match capability {
                    DeployerTypes::VirtualMachine => {
                        self.machiners.insert(
                            deployer.deployer_name.to_owned(),
                            deployer.deployer_ip.to_owned(),
                        );
                    }
                    DeployerTypes::Switch => {
                        self.switchers.insert(
                            deployer.deployer_name.to_owned(),
                            deployer.deployer_ip.to_owned(),
                        );
                    }
                    DeployerTypes::Template => {
                        self.templaters.insert(
                            deployer.deployer_name.to_owned(),
                            deployer.deployer_ip.to_owned(),
                        );
                    }
                });
        }
    }
}

#[derive(Debug, Default, Clone, Serialize, MessageResponse)]
pub struct DeployerGroups(pub HashMap<String, DeployerGroup>);

impl DeployerGroups {
    pub fn new() -> Self {
        DeployerGroups(HashMap::new())
    }

    pub async fn start_all(
        deployer_grouper_address: Addr<DeployerGroups>,
        deployment_manager_address: Addr<DeploymentManager>,
    ) -> Result<()> {
        let deployer_groups = deployer_grouper_address.send(GetDeployerGroups).await?;
        let deployment_actors = join_all(
            deployer_groups
                .0
                .into_iter()
                .map(|(name, deployer_group)| async move { (name, deployer_group.start().await) }),
        )
        .await;

        let mut deployment_groups: DeploymentGroupMap = HashMap::new();
        deployment_actors
            .iter()
            .for_each(|(name, deployment_group)| {
                deployment_groups.insert(name.to_string(), deployment_group.clone());
            });
        deployment_manager_address
            .send(AddDeploymentGroups(deployment_groups))
            .await?;

        Ok(())
    }

    pub fn initialize_with_group_names(
        deployment_groups: &HashMap<String, Vec<String>>,
    ) -> DeployerGroups {
        let mut deployer_groups = DeployerGroups::default();
        deployment_groups.iter().for_each(|deployer_group| {
            deployer_groups
                .0
                .insert(deployer_group.0.to_owned(), DeployerGroup::default());
        });
        deployer_groups
    }
}

impl Actor for DeployerGroups {
    type Context = Context<Self>;
}

impl Handler<AddDeployerGroups> for DeployerGroups {
    type Result = ();
    fn handle(&mut self, msg: AddDeployerGroups, _: &mut Context<Self>) -> Self::Result {
        self.0.extend(msg.0 .0);
    }
}

impl Handler<GetDeployerGroups> for DeployerGroups {
    type Result = DeployerGroups;
    fn handle(&mut self, _: GetDeployerGroups, _: &mut Context<Self>) -> Self::Result {
        self.to_owned()
    }
}

fn filter_capable_deployers(capability_results: Vec<Result<Deployer>>) -> Vec<Deployer> {
    capability_results
        .into_iter()
        .filter_map(|result| {
            result
                .map_err(|error| error!("Error getting deployer capability: {}", error))
                .ok()
        })
        .collect()
}

pub async fn get_deployer_capabilities(
    deployers: HashMap<String, String>,
) -> Result<Vec<Deployer>> {
    let capability_results = join_all(deployers.into_iter().map(|deployer| async move {
        let deployer_client = NodeClient::new(deployer.1.to_owned()).await?.start();
        let capabilities = deployer_client.send(GetCapabilities).await??;
        Ok(Deployer {
            deployer_name: deployer.0.to_owned(),
            deployer_ip: deployer.1.to_owned(),
            capabilities: Some(capabilities),
        })
    }))
    .await;
    let verified_deployers = filter_capable_deployers(capability_results);
    if verified_deployers.is_empty() {
        error!("No deployers found with capabilities");
        return Err(anyhow!("No deployers found with capabilities"));
    }
    Ok(verified_deployers)
}
