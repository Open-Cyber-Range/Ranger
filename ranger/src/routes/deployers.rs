use crate::AppState;
use actix_web::web::Data;
use actix_web::{get, Error, HttpResponse};
use anyhow::Result;

#[get("deployer")]
pub async fn get_deployers(app_state: Data<AppState>) -> Result<HttpResponse, Error> {
    let deployment_groups = app_state.configuration.deployment_groups.clone();

    Ok(HttpResponse::Ok()
        .content_type("application/json")
        .json(deployment_groups))
}
