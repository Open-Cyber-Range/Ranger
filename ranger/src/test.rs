
#[cfg(test)]
mod tests {


    fn setup_test_server() -> Result<(String, Data<AppState>)> {
        let app_state = create_test_app_state();
    }

    pub fn create_test_app_state() -> Data<AppState> {
        let configuration = create_test_configuration();
        let node_client_address = create_test_node_client_address();
        let database_address = create_test_database_address();
        let deployer_address = create_test_deployer_address();
        let app_state = AppState {
            database_address,
            deployer_address,
        };
        Data::new(app_state)
    }
}