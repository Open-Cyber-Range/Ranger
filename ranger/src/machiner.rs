use crate::deployers::{DeployerGroups, GetDeployerGroups};
use crate::node::{CreateNode, NodeClient};
use actix::{
    Actor, ActorFutureExt, Addr, Context, Handler, Message, ResponseActFuture, WrapFuture,
};
use anyhow::{anyhow, Ok, Result};
use futures::future::try_join_all;
use log::info;
use ranger_grpc::{DeploymentParameters, Node, NodeDeployment, NodeIdentifier, NodeType};
use sdl_parser::Scenario;
use std::collections::HashMap;
use uuid::Uuid;

#[derive(Message, Debug)]
#[rtype(result = "Result<Uuid>")]
pub struct CreateDeployment(pub(crate) Scenario);

#[derive(Message, Debug)]
#[rtype(result = "()")]
pub struct DeleteDeployment(pub(crate) String);

#[derive(Debug, Clone)]
pub struct DeploymentManager {
    nodes: HashMap<String, HashMap<Uuid, Vec<(NodeIdentifier, String)>>>,
    node_client_address: Addr<NodeClient>,
}

impl DeploymentManager {
    pub async fn new(
        deployer_actor_address: Addr<DeployerGroups>,
        deployment_group_name: String,
    ) -> Result<Self> {
        let validated_deployer_groups = deployer_actor_address.send(GetDeployerGroups).await?;
        if let Some(deployer_group) = validated_deployer_groups.0.get(&deployment_group_name) {
            if let Some(machiner) = deployer_group.machiners.values().next() {
                return Ok(DeploymentManager {
                    nodes: HashMap::new(),
                    node_client_address: NodeClient::new(machiner.to_string()).await?.start(),
                });
            }
            return Err(anyhow!("No machiners found"));
        }
        Err(anyhow!(
            "Deployer group named {deployment_group_name} not found"
        ))
    }
}

impl Actor for DeploymentManager {
    type Context = Context<Self>;
}

impl Handler<CreateDeployment> for DeploymentManager {
    type Result = ResponseActFuture<Self, Result<Uuid>>;

    fn handle(&mut self, msg: CreateDeployment, _: &mut Context<Self>) -> Self::Result {
        let scenario = msg.0;
        let client_address = self.node_client_address.clone();
        let scenario_name = scenario.name.clone();
        Box::pin(
            async move {
                let deployment_id = Uuid::new_v4();
                let exercise_name = format!("{}-{}", scenario_name, deployment_id);
                let node_ids: Vec<(NodeIdentifier, String)> =
                    if let Some(infrastructure) = scenario.infrastructure {
                        try_join_all(infrastructure.into_iter().map(|(node_name, node)| async {
                            let node_id = client_address
                                .send(CreateNode(NodeDeployment {
                                    parameters: Some(DeploymentParameters {
                                        name: node_name.clone(),
                                        exercise_name: exercise_name.clone(),
                                        template_name: node
                                            .source
                                            .ok_or_else(|| anyhow!("Source is missing"))?
                                            .template
                                            .ok_or_else(|| anyhow!("Template is missing"))?,
                                    }),
                                    node: Some(Node {
                                        identifier: Some(NodeIdentifier {
                                            identifier: None,
                                            node_type: NodeType::Vm.into(),
                                        }),
                                        configuration: None,
                                    }),
                                }))
                                .await??;
                            Ok::<(NodeIdentifier, String)>((node_id, node_name))
                        }))
                        .await?
                    } else {
                        Vec::new()
                    };
                Ok((deployment_id, node_ids))
            }
            .into_actor(self)
            .map(move |result, act, _| {
                if let core::result::Result::Ok((deployment_id, node_ids)) = result {
                    info!(
                        "Successful deployment id: {deployment_id}, deployed {} nodes",
                        &node_ids.len()
                    );
                    act.nodes
                        .entry(scenario.name)
                        .or_insert_with(HashMap::new)
                        .insert(deployment_id, node_ids);
                    Ok(deployment_id)
                } else {
                    Err(anyhow!(
                        "Deployment failed: {:?}",
                        result
                            .err()
                            .unwrap_or_else(|| anyhow!("DeploymentManager error"))
                    ))
                }
            }),
        )
    }
}
