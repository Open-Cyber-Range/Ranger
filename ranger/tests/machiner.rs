mod common;

#[cfg(test)]
mod tests {
    use crate::common::create_mock_vmware_server;
    use actix_web::test;
    use anyhow::{anyhow, Result};
    use ranger::deployers::DeployerGroup;
    use ranger::machiner::NodeDeploymentTrait;
    use ranger_grpc::NodeDeployment;
    use sdl_parser::test::TEST_SCHEMA;

    #[test]
    async fn verify_vm_node_deployment_struct() -> Result<()> {
        let mut nodes = TEST_SCHEMA
            .scenario
            .nodes
            .clone()
            .ok_or_else(|| anyhow!("TEST: Nodes list is missing"))?;
        let node = nodes
            .remove_entry("win10")
            .ok_or_else(|| anyhow!("TEST: Node not found"))?;
        let deployment =
            NodeDeployment::default().initialize_vm(node, "test-exercise".to_string())?;
        insta::assert_debug_snapshot!(deployment);
        Ok(())
    }

    #[actix_web::test]
    pub async fn mock_deploy_vms() -> Result<()> {
        let nodes = TEST_SCHEMA
            .scenario
            .nodes
            .clone()
            .ok_or_else(|| anyhow!("TEST: Nodes list is missing"))?;

        let mock_node_server = create_mock_vmware_server().run_node_server()?;
        let mut deployer_group = DeployerGroup::default();
        deployer_group.machiners.insert(
            "mock-server".to_string(),
            format!("http://{}", mock_node_server),
        );
        let deployment_group = deployer_group.start().await;

        let mut node_identifiers = deployment_group.deploy_vms(nodes, "test-exercise").await?;
        node_identifiers.sort_by_key(|node_name| node_name.1.clone());
        insta::assert_debug_snapshot!(node_identifiers);
        Ok(())
    }
}
