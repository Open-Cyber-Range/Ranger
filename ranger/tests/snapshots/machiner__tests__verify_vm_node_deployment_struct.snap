---
source: ranger/tests/machiner.rs
expression: deployment
---
NodeDeployment {
    parameters: Some(
        DeploymentParameters {
            name: "win10",
            exercise_name: "test-exercise",
            template_id: "windows10",
        },
    ),
    node: Some(
        Node {
            identifier: Some(
                NodeIdentifier {
                    identifier: None,
                    node_type: Vm,
                },
            ),
            configuration: None,
        },
    ),
}
